File: test_agents_api_medical.py
#!/usr/bin/env python3
"""
Agent Studio API Test Suite for asgard-medical Tenant

Tests agent management and chat functionality:
- List all agents
- Get agent templates
- Retrieve specific agent config
- Chat with an agent
- Conversation management
- Agent routing

Usage:
    python test_agents_api_medical.py
    MIMIR_URL=http://localhost:3002 python test_agents_api_medical.py
"""

import os
import sys
import json
import requests
from datetime import datetime
from typing import Dict, Any, Optional
from urllib.parse import urljoin

# Configuration
BASE_URL = os.environ.get('MIMIR_URL', 'http://localhost:3002')
TENANT_ID = 'asgard-medical'
TEST_TIMEOUT = 30

# ANSI Color codes
class Colors:
    RESET = '\033[0m'
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'
    MAGENTA = '\033[35m'

class AgentTestSuite:
    def __init__(self):
        self.tests_passed = 0
        self.tests_failed = 0
        self.session = requests.Session()
        self.agents = []

    def log(self, message: str, level: str = 'info') -> None:
        """Log message with timestamp and color."""
        timestamp = datetime.now().isoformat()

        color_map = {
            'pass': Colors.GREEN,
            'fail': Colors.RED,
            'warn': Colors.YELLOW,
            'section': Colors.CYAN,
            'info': Colors.BLUE,
            'debug': Colors.MAGENTA
        }

        color = color_map.get(level, Colors.BLUE)
        print(f"{color}[{timestamp}]{Colors.RESET} {message}")

    def make_request(
        self,
        method: str,
        path: str,
        data: Optional[Dict] = None,
        expected_status: int = 200
    ) -> tuple[Optional[Dict], bool]:
        """Make HTTP request and return (response_data, success)."""
        try:
            url = urljoin(BASE_URL, path)

            if method == 'GET':
                resp = self.session.get(url, timeout=TEST_TIMEOUT)
            elif method == 'POST':
                resp = self.session.post(
                    url,
                    json=data,
                    headers={'Content-Type': 'application/json'},
                    timeout=TEST_TIMEOUT
                )
            else:
                return None, False

            success = resp.status_code == expected_status

            try:
                response_data = resp.json()
            except:
                response_data = {'raw': resp.text}

            return response_data, success
        except Exception as e:
            self.log(f"Request error: {str(e)}", 'fail')
            return None, False

    def test_list_agents(self) -> None:
        """Test: List all available agents for the tenant."""
        path = f'/api/v1/agents'
        data, success = self.make_request('GET', path, expected_status=200)

        if success and isinstance(data, list):
            self.agents = data
            agent_count = len(data)

            self.log(f"✓ Found {agent_count} agents", 'pass')

            # Display agent details
            print(f"\n  {'Agent ID':<25} {'Name':<30} {'Specialty':<20} {'Model':<30}")
            print(f"  {'-'*25} {'-'*30} {'-'*20} {'-'*30}")

            for agent in data:
                agent_id = agent.get('id', 'N/A')[:25]
                name = agent.get('name', 'Unnamed')[:30]
                specialty = agent.get('specialty', 'General')[:20]
                model = agent.get('model_id', 'N/A')[:30]

                print(f"  {agent_id:<25} {name:<30} {specialty:<20} {model:<30}")

            self.tests_passed += 1
        else:
            self.log(f"✗ Failed to list agents", 'fail')
            self.tests_failed += 1

    def test_get_agent_templates(self) -> None:
        """Test: Get available agent templates."""
        path = f'/api/v1/agents/templates'
        data, success = self.make_request('GET', path, expected_status=200)

        if success and isinstance(data, list):
            template_count = len(data)
            self.log(f"✓ Found {template_count} agent templates", 'pass')

            # Display first few templates
            for template in data[:3]:
                template_name = template.get('name', 'Unknown')
                template_desc = template.get('description', '')[:60]
                print(f"  - {template_name}: {template_desc}...")

            self.tests_passed += 1
        else:
            self.log(f"✗ Failed to get templates", 'fail')
            self.tests_failed += 1

    def test_get_specific_agent(self) -> None:
        """Test: Get details of a specific agent."""
        if not self.agents:
            self.log("⊘ Skipping: No agents available", 'warn')
            return

        # Test first agent
        agent_id = self.agents[0].get('id')
        path = f'/api/v1/agents/{agent_id}'
        data, success = self.make_request('GET', path, expected_status=200)

        if success and data:
            name = data.get('name', 'Unknown')
            specialty = data.get('specialty', 'General')
            use_rag = data.get('use_rag', False)
            temperature = data.get('temperature', 0)

            self.log(
                f"✓ Retrieved agent '{name}' (RAG: {use_rag}, Temp: {temperature})",
                'pass'
            )
            self.tests_passed += 1
        else:
            self.log(f"✗ Failed to get agent details", 'fail')
            self.tests_failed += 1

    def test_agent_chat(self) -> None:
        """Test: Send a chat message to an agent."""
        if not self.agents:
            self.log("⊘ Skipping: No agents available", 'warn')
            return

        # Use first agent for chat
        agent_id = self.agents[0].get('id')
        path = f'/api/v1/agents/{agent_id}/chat'

        chat_request = {
            "message": "What are common symptoms of respiratory infections?",
            "model": "auto"  # Use agent's default model
        }

        data, success = self.make_request('POST', path, data=chat_request, expected_status=200)

        if success and data:
            response = data.get('response', '')
            conversation_id = data.get('conversation_id', 'N/A')

            if response:
                preview = response[:100] + "..." if len(response) > 100 else response
                self.log(
                    f"✓ Chat successful (Conv: {conversation_id})",
                    'pass'
                )
                print(f"\n  Response Preview:\n  \"{preview}\"\n")
                self.tests_passed += 1
            else:
                self.log(f"✗ No response from agent", 'fail')
                self.tests_failed += 1
        else:
            self.log(f"✗ Chat request failed", 'fail')
            self.tests_failed += 1

    def test_agent_conversations(self) -> None:
        """Test: List conversations for an agent."""
        if not self.agents:
            self.log("⊘ Skipping: No agents available", 'warn')
            return

        agent_id = self.agents[0].get('id')
        path = f'/api/v1/agents/{agent_id}/conversations'

        data, success = self.make_request('GET', path, expected_status=200)

        if success and isinstance(data, list):
            conv_count = len(data)
            self.log(
                f"✓ Retrieved {conv_count} conversations for agent",
                'pass'
            )

            # Show first few conversations
            for conv in data[:3]:
                conv_id = conv.get('id', 'N/A')
                created = conv.get('created_at', 'N/A')
                print(f"  - Conversation {conv_id} (Created: {created})")

            self.tests_passed += 1
        else:
            self.log(f"✗ Failed to list conversations", 'fail')
            self.tests_failed += 1

    def test_agent_routing(self) -> None:
        """Test: Route a question to appropriate specialist agent."""
        path = f'/api/v1/agents/route'

        route_request = {
            "question": "I have chest pain and difficulty breathing. What specialist should I see?",
            "tenant_id": TENANT_ID
        }

        data, success = self.make_request('POST', path, data=route_request, expected_status=200)

        if success and data:
            selected_agent = data.get('selected_agent', {})
            agent_name = selected_agent.get('name', 'Unknown')
            specialty = selected_agent.get('specialty', 'Unknown')
            confidence = data.get('confidence', 0)

            self.log(
                f"✓ Routed to specialist: '{agent_name}' ({specialty}) - Confidence: {confidence:.2f}",
                'pass'
            )
            self.tests_passed += 1
        else:
            self.log(f"✗ Agent routing failed", 'fail')
            self.tests_failed += 1

    def test_agent_list_query(self) -> None:
        """Test: List agents with query filters."""
        path = f'/api/v1/agents?specialty=cardiology&limit=5'

        data, success = self.make_request('GET', path, expected_status=200)

        if success and isinstance(data, list):
            self.log(
                f"✓ Filtered agents by specialty: {len(data)} results",
                'pass'
            )
            self.tests_passed += 1
        else:
            self.log(f"✗ Failed to filter agents", 'fail')
            self.tests_failed += 1

    def run_all_tests(self) -> None:
        """Run complete test suite."""
        self.log('=== Agent Studio API Test Suite ===', 'section')
        self.log(f'Base URL: {BASE_URL}', 'info')
        self.log(f'Tenant: {TENANT_ID}\n', 'info')

        # Phase 1: Agent Discovery
        self.log('Phase 1: Agent Discovery', 'section')
        self.test_list_agents()
        self.test_get_agent_templates()

        # Phase 2: Agent Management
        self.log('\nPhase 2: Agent Management', 'section')
        self.test_get_specific_agent()

        # Phase 3: Agent Chat
        self.log('\nPhase 3: Agent Chat & Conversations', 'section')
        self.test_agent_chat()
        self.test_agent_conversations()

        # Phase 4: Advanced Features
        self.log('\nPhase 4: Advanced Features', 'section')
        self.test_agent_routing()
        self.test_agent_list_query()

        # Summary
        self.log('\n=== Test Summary ===', 'section')
        total = self.tests_passed + self.tests_failed
        percentage = (self.tests_passed / total * 100) if total > 0 else 0

        summary_color = (
            Colors.GREEN if self.tests_failed == 0 else
            Colors.YELLOW if self.tests_failed <= 2 else
            Colors.RED
        )

        print(f"""
{summary_color}
Passed:  {self.tests_passed}/{total}
Failed:  {self.tests_failed}/{total}
Success: {percentage:.1f}%
{Colors.RESET}""")

        return 0 if self.tests_failed == 0 else 1

def main():
    """Main entry point."""
    try:
        suite = AgentTestSuite()
        exit_code = suite.run_all_tests()
        sys.exit(exit_code)
    except KeyboardInterrupt:
        print("\n\nTest run interrupted")
        sys.exit(1)
    except Exception as e:
        print(f"Fatal error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()

File: test_agents_api_medical.js
#!/usr/bin/env node
/**
 * Agent Studio API Test Suite for asgard-medical Tenant
 *
 * Tests agent management and chat functionality:
 * - List all agents
 * - Get agent templates
 * - Retrieve specific agent config
 * - Chat with an agent
 * - Conversation management
 * - Agent routing
 *
 * Usage:
 *   node test_agents_api_medical.js
 *   MIMIR_URL=http://localhost:3002 node test_agents_api_medical.js
 */

const BASE_URL = process.env.MIMIR_URL || 'http://localhost:3002';
const TENANT_ID = 'asgard-medical';
const TEST_TIMEOUT = 30000;

let testsPassed = 0;
let testsFailed = 0;
let agents = [];

// Color codes for terminal output
const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m',
  magenta: '\x1b[35m'
};

async function log(message, type = 'info') {
  const timestamp = new Date().toISOString();
  const color = type === 'pass' ? colors.green :
                type === 'fail' ? colors.red :
                type === 'warn' ? colors.yellow :
                type === 'section' ? colors.cyan :
                type === 'debug' ? colors.magenta : colors.blue;
  console.log(`${color}[${timestamp}]${colors.reset} ${message}`);
}

async function makeRequest(method, path, data = null, expectedStatus = 200) {
  try {
    const options = {
      method,
      headers: { 'Content-Type': 'application/json' },
      timeout: TEST_TIMEOUT
    };

    if (data) {
      options.body = JSON.stringify(data);
    }

    const res = await fetch(`${BASE_URL}${path}`, options);
    const result = await res.json().catch(() => ({ raw: res.statusText }));

    return [result, res.status === expectedStatus];
  } catch (error) {
    await log(`Request error: ${error.message}`, 'fail');
    return [null, false];
  }
}

async function testListAgents() {
  const [data, success] = await makeRequest('GET', '/api/v1/agents');

  if (success && Array.isArray(data)) {
    agents = data;
    await log(`✓ Found ${agents.length} agents`, 'pass');

    console.log(`\n  ${'Agent ID':<25} ${'Name':<30} ${'Specialty':<20} ${'Model':<30}`);
    console.log(`  ${'-'.repeat(25)} ${'-'.repeat(30)} ${'-'.repeat(20)} ${'-'.repeat(30)}`);

    agents.forEach(agent => {
      const agentId = (agent.id || 'N/A').substring(0, 25).padEnd(25);
      const name = (agent.name || 'Unnamed').substring(0, 30).padEnd(30);
      const specialty = (agent.specialty || 'General').substring(0, 20).padEnd(20);
      const model = (agent.model_id || 'N/A').substring(0, 30).padEnd(30);
      console.log(`  ${agentId} ${name} ${specialty} ${model}`);
    });

    testsPassed++;
  } else {
    await log('✗ Failed to list agents', 'fail');
    testsFailed++;
  }
}

async function testGetAgentTemplates() {
  const [data, success] = await makeRequest('GET', '/api/v1/agents/templates');

  if (success && Array.isArray(data)) {
    await log(`✓ Found ${data.length} agent templates`, 'pass');

    data.slice(0, 3).forEach(template => {
      const desc = (template.description || '').substring(0, 60);
      console.log(`  - ${template.name}: ${desc}...`);
    });

    testsPassed++;
  } else {
    await log('✗ Failed to get templates', 'fail');
    testsFailed++;
  }
}

async function testGetSpecificAgent() {
  if (agents.length === 0) {
    await log('⊘ Skipping: No agents available', 'warn');
    return;
  }

  const agentId = agents[0].id;
  const [data, success] = await makeRequest('GET', `/api/v1/agents/${agentId}`);

  if (success && data) {
    const name = data.name || 'Unknown';
    const specialty = data.specialty || 'General';
    const useRag = data.use_rag || false;
    const temperature = data.temperature || 0;

    await log(
      `✓ Retrieved agent '${name}' (RAG: ${useRag}, Temp: ${temperature})`,
      'pass'
    );
    testsPassed++;
  } else {
    await log('✗ Failed to get agent details', 'fail');
    testsFailed++;
  }
}

async function testAgentChat() {
  if (agents.length === 0) {
    await log('⊘ Skipping: No agents available', 'warn');
    return;
  }

  const agentId = agents[0].id;
  const chatRequest = {
    message: 'What are common symptoms of respiratory infections?',
    model: 'auto'
  };

  const [data, success] = await makeRequest(
    'POST',
    `/api/v1/agents/${agentId}/chat`,
    chatRequest
  );

  if (success && data) {
    const response = data.response || '';
    const conversationId = data.conversation_id || 'N/A';

    if (response) {
      const preview = response.length > 100 ? response.substring(0, 100) + '...' : response;
      await log(
        `✓ Chat successful (Conv: ${conversationId})`,
        'pass'
      );
      console.log(`\n  Response Preview:\n  "${preview}"\n`);
      testsPassed++;
    } else {
      await log('✗ No response from agent', 'fail');
      testsFailed++;
    }
  } else {
    await log('✗ Chat request failed', 'fail');
    testsFailed++;
  }
}

async function testAgentConversations() {
  if (agents.length === 0) {
    await log('⊘ Skipping: No agents available', 'warn');
    return;
  }

  const agentId = agents[0].id;
  const [data, success] = await makeRequest('GET', `/api/v1/agents/${agentId}/conversations`);

  if (success && Array.isArray(data)) {
    await log(
      `✓ Retrieved ${data.length} conversations for agent`,
      'pass'
    );

    data.slice(0, 3).forEach(conv => {
      const convId = conv.id || 'N/A';
      const created = conv.created_at || 'N/A';
      console.log(`  - Conversation ${convId} (Created: ${created})`);
    });

    testsPassed++;
  } else {
    await log('✗ Failed to list conversations', 'fail');
    testsFailed++;
  }
}

async function testAgentRouting() {
  const routeRequest = {
    question: 'I have chest pain and difficulty breathing. What specialist should I see?',
    tenant_id: TENANT_ID
  };

  const [data, success] = await makeRequest(
    'POST',
    '/api/v1/agents/route',
    routeRequest
  );

  if (success && data) {
    const selectedAgent = data.selected_agent || {};
    const agentName = selectedAgent.name || 'Unknown';
    const specialty = selectedAgent.specialty || 'Unknown';
    const confidence = data.confidence || 0;

    await log(
      `✓ Routed to specialist: '${agentName}' (${specialty}) - Confidence: ${confidence.toFixed(2)}`,
      'pass'
    );
    testsPassed++;
  } else {
    await log('✗ Agent routing failed', 'fail');
    testsFailed++;
  }
}

async function testAgentListQuery() {
  const [data, success] = await makeRequest(
    'GET',
    '/api/v1/agents?specialty=cardiology&limit=5'
  );

  if (success && Array.isArray(data)) {
    await log(
      `✓ Filtered agents by specialty: ${data.length} results`,
      'pass'
    );
    testsPassed++;
  } else {
    await log('✗ Failed to filter agents', 'fail');
    testsFailed++;
  }
}

async function runAllTests() {
  await log('=== Agent Studio API Test Suite ===', 'section');
  await log(`Base URL: ${BASE_URL}`, 'info');
  await log(`Tenant: ${TENANT_ID}\n`, 'info');

  // Phase 1: Agent Discovery
  await log('Phase 1: Agent Discovery', 'section');
  await testListAgents();
  await testGetAgentTemplates();

  // Phase 2: Agent Management
  await log('\nPhase 2: Agent Management', 'section');
  await testGetSpecificAgent();

  // Phase 3: Agent Chat
  await log('\nPhase 3: Agent Chat & Conversations', 'section');
  await testAgentChat();
  await testAgentConversations();

  // Phase 4: Advanced Features
  await log('\nPhase 4: Advanced Features', 'section');
  await testAgentRouting();
  await testAgentListQuery();

  // Summary
  await log('\n=== Test Summary ===', 'section');
  const total = testsPassed + testsFailed;
  const percentage = total > 0 ? Math.round((testsPassed / total) * 100) : 0;

  const summaryColor = testsFailed === 0 ? colors.green :
                       testsFailed <= 2 ? colors.yellow : colors.red;

  console.log(`${summaryColor}
Passed:  ${testsPassed}/${total}
Failed:  ${testsFailed}/${total}
Success: ${percentage}%
${colors.reset}`);

  process.exit(testsFailed > 0 ? 1 : 0);
}

// Handle graceful shutdown
process.on('SIGINT', async () => {
  await log('\n\nTest run interrupted', 'warn');
  process.exit(1);
});

// Run tests
runAllTests().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});
